### PR TITLE
Fix: Bug with neuron appearing

### PR DESCRIPTION
**Pull request by lmuntaner**
# Motivation

A bug was raised in the forum where an SNS neuron was appearing even though the user transferred ownership of such a neuron.

The issue came from the background task that checks neuron subaccounts. The background task didn't consider whether the user has still some permission with the neuron and refreshes it and adds it to the store.

The fix is to try to query the neuron from the canister directly and then check the permissions before continuing with refresh or not refresh.

# Changes

* Add code in the main `checkNeuronsSubaccounts` to fetch the neuron when it wasn't passed and compare the permissions with the user's identity.

# Tests

* Add two tests regarding the bug. One simulating the bug and another one with an edge case to confirm that if a neuron without permissions is found, it doesn't break the loop of claiming neurons.
* Fix the tests in sns-neurons that were broken due to expecting the user to be the controller of the neuron.
* Add a missing test when a neuron that wasn't created in NNS Dapp needed to be refreshed.

# Todos

- [x] Add entry to changelog (if necessary).

### CODE REVIEW

**Comment by lmuntaner** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #4484** <a href="https://app.graphite.dev/github/pr/dfinity/nns-dapp/4484?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/dfinity/nns-dapp/4484?utm_source=stack-comment).

File: frontend/src/lib/services/sns-neurons-check-balances.services.ts [hunk 1]
```diff
@@ -185,9 +186,29 @@ const checkNeuronsSubaccounts = async ({
         certified: false,
         identity,
       });
-      const neuron = neurons.find(findNeuronBySubaccount(subaccount));
-      const neuronNotFound = neuron === undefined;
       const positiveBalance = currentBalance >= neuronMinimumStake;
```
Comment (lmuntaner) [inline frontend/src/lib/services/sns-neurons-check-balances.services.ts:189]: This function is big and could be refactored further.

I might try to do that in another PR, but I believe that a PR with only fixing it and having the diff clearly identifiable is worth it.

**Comment by lmuntaner** [comment]
@dskloetd please review

File: frontend/src/lib/services/sns-neurons-check-balances.services.ts (hunk 1 above)
Comment (dskloetd) [reply frontend/src/lib/services/sns-neurons-check-balances.services.ts:189]: The whole added code could be a separate function without refactoring the existing code.
Then the existing function would only be slightly longer instead of much longer.
```
const hasUserSomePermission = userHasSomePermission(...);
```

File: frontend/src/lib/services/sns-neurons-check-balances.services.ts [hunk 2]
```diff
@@ -225,6 +246,9 @@ const checkNeuronsSubaccounts = async ({
       console.error(error);
     }
   }
+  // Not all neurons that the user has some permission need to be creeted in NNS Dapp.
```
Comment (dskloetd) [inline frontend/src/lib/services/sns-neurons-check-balances.services.ts:249]: typo: creeted

**Review by dskloetd — COMMENTED** [review]
I'll review the tests tomorrow.

File: frontend/src/lib/services/sns-neurons-check-balances.services.ts [hunk 3]
```diff
@@ -110,9 +111,56 @@ const claimAndLoadNeuron = async ({
...
+      getSnsNeuronIdAsHexString(neuron) === subaccountToHexString(subaccount)
+  );
+  // At this point we don't know whether it's an unclaimed or transferred neuron.
+  if (isNullish(maybeNeuron) && positiveBalance) {
+    const neuronId: SnsNeuronId = { id: subaccount };
+    maybeNeuron = await querySnsNeuron({
+      identity: new AnonymousIdentity(),
+      rootCanisterId,
+      neuronId,
+      // No need to check with update call, worst case, a neuron will appear in the UI that shouldn't.
```
Comment (dskloetd) [inline frontend/src/lib/services/sns-neurons-check-balances.services.ts:136]: Isn't it also possible that no neuron appears while one should appear?

**Review by dskloetd — APPROVED** [review]

File: frontend/src/lib/services/sns-neurons-check-balances.services.ts (hunk 1 above)
Comment (lmuntaner) [reply frontend/src/lib/services/sns-neurons-check-balances.services.ts:236]: done

File: frontend/src/lib/services/sns-neurons-check-balances.services.ts (hunk 3 above)
Comment (lmuntaner) [reply frontend/src/lib/services/sns-neurons-check-balances.services.ts:136]: Yes, true. In that case an unnecessary call to refresh would be done.

Let me improve the comment.

File: frontend/src/lib/services/sns-neurons-check-balances.services.ts (hunk 2 above)
Comment (lmuntaner) [reply frontend/src/lib/services/sns-neurons-check-balances.services.ts]: done, thanks!
